File: src/main/java/io/github/m/Stdio.java
package io.github.m;

import java.io.IOException;

public final class Stdio {
    @MField(name = "stdout")
    public static final Value stdout = new Value.Impl1("stdout", arg ->
            new Process.Impl("(stdout char)", () -> {
                System.out.write(Char.from(arg).value);
                return List.NIL;
            }));

    @MField(name = "stderr")
    public static final Value stderr = new Value.Impl1("stderr", arg ->
            new Process.Impl("(stderr char)", () -> {
                System.err.write(Char.from(arg).value);
                return List.NIL;
            }));

    @MField(name = "stdin")
    public static final Value stdin = new Process.Impl("stdin", () -> {
        System.out.flush();
        System.err.flush();
        try {
            return Char.valueOf((char) System.in.read());
        } catch (IOException e) {
            return Error.wrap(e);
        }
    });
}

File: src/main/kotlin/io/github/m/Stdio.kt
package io.github.m

object Stdio {
    object Stdout : Value {
        override fun invoke(arg: Value) = Process { System.out.write(Char.from(arg).char.toInt()); List.NIL }
    }

    object Stderr : Value {
        override fun invoke(arg: Value) = Process { System.err.write(Char.from(arg).char.toInt()); List.NIL }
    }

    object Stdin : Process {
        override fun run() = run {
            System.out.flush()
            Char(System.`in`.read().toChar())
        }
    }

    /**
     * M stdio definitions.
     */
    @Suppress("unused")
    object Definitions {
        @MField(name = "stdout")
        @JvmField
        val stdout: Value = Stdout

        @MField(name = "stderr")
        @JvmField
        val stderr: Value = Stderr

        @MField(name = "stdin")
        @JvmField
        val stdin: Value = Stdin
    }
}